### PR TITLE
feat: NotificationSource to receive data from G7 and Omnipod 5

### PR DESCRIPTION
**Pull request by rgodha24**
adds a source that grabs data from dexcom g7 and omnipod 5 notifications

### CODE REVIEW

File: common/src/main/java/de/michelinside/glucodatahandler/common/receiver/NamedBroadcastReceiver.kt [hunk 1]
```diff
@@ -5,8 +5,7 @@ import android.content.Context
 import android.content.Intent
 import de.michelinside.glucodatahandler.common.GlucoDataService
 
-abstract class ReceiverBase: BroadcastReceiver() {
-    abstract fun getName(): String
+abstract class NamedBroadcastReceiver: BroadcastReceiver(), NamedReceiver {
```
Comment (pachi81) [inline common/src/main/java/de/michelinside/glucodatahandler/common/receiver/NamedBroadcastReceiver.kt:8]: Renaming ReceiverBase will cause more changes than needed and may cause conflicts. 
I think, this should be done during refactoring, not for a new feature...
But it is ok for me, as there are not so many related changes...

File: common/src/main/java/de/michelinside/glucodatahandler/common/receiver/NotificationReceiver.kt [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+            Log.d(LOG_ID, "Found ${textViews.size} text views")
+
+            var matches = 0
+            var glucoseValue = 0
+
+            // Examine each TextView
+            for (textView in textViews) {
+                try {
+                    val bg = textView.text?.toString()?.toIntOrNull() ?: 0
+                    if (bg > 0) {
```
Comment (pachi81) [inline common/src/main/java/de/michelinside/glucodatahandler/common/receiver/NotificationReceiver.kt:66]: Please check the value using GlucoDataUtils.isGlucoseValid(bg) for matching

File: common/src/main/java/de/michelinside/glucodatahandler/common/receiver/NotificationReceiver.kt [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+            Log.d(LOG_ID, "Found ${textViews.size} text views")
+
+            var matches = 0
+            var glucoseValue = 0
+
+            // Examine each TextView
+            for (textView in textViews) {
+                try {
+                    val bg = textView.text?.toString()?.toIntOrNull() ?: 0
```
Comment (pachi81) [inline common/src/main/java/de/michelinside/glucodatahandler/common/receiver/NotificationReceiver.kt:65]: Better using float for mmol/l

File: common/src/main/java/de/michelinside/glucodatahandler/common/receiver/NotificationReceiver.kt [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        }
+    }
+
+    private fun handleGlucoseValue(glucoseValue: Int) {
+        Log.d(LOG_ID, "Extracted glucose value: $glucoseValue")
+        // TODO: g7 notifications update weirdly often. deduplication is needed
+        if (GlucoDataUtils.isGlucoseValid(glucoseValue)) {
+            val glucoExtras = Bundle()
+            glucoExtras.putLong(ReceiveData.TIME, System.currentTimeMillis())
+            glucoExtras.putInt(ReceiveData.MGDL,glucoseValue)
```
Comment (pachi81) [inline common/src/main/java/de/michelinside/glucodatahandler/common/receiver/NotificationReceiver.kt:115]: glucoseValue may also be a mmol/ value, so you should check using GlucoDataUtils.isMmolValue().
If it is you should also set the GLUCOSECUSTOM like here:
https://github.com/pachi81/GlucoDataHandler/blob/679aa3241dc39cc457148600c2cd71e59494ec57/common/src/main/java/de/michelinside/glucodatahandler/common/tasks/NightscoutSourceTask.kt#L192

**Review by pachi81 — CHANGES_REQUESTED** [review]
First of all, thank you!!!!!!
It seems, you have understood my code :-D
I already planned this feature for the future, so this will save me some time. 

One remark, I won´t publish this feature until I have implemented the graph including historical data, as this will also include some basic trend calculation, which is not really possible without historical data.
I already have this problem with Diabox source and the first users complain about, even if I have mentioned in docu, that there is no trend support, yet...

One more question, what about an packagename filter? For only handling notification from this package!?

I have also thought about making a small app, only doing this stuff and show all parts of the notification in main view. So users from other apps, like CampAPS can send me screenshots of the content extracted from the notification. What do you think?

File: common/src/main/java/de/michelinside/glucodatahandler/common/receiver/NotificationReceiver.kt (hunk 4 above)
Comment (pachi81) [reply common/src/main/java/de/michelinside/glucodatahandler/common/receiver/NotificationReceiver.kt:115]: I added you as contributor, so it may be easier to create PRs.
https://github.com/pachi81/GlucoDataHandler/invitations

**Comment by pachi81** [comment]
I added you as contributor, so it may be easier to create PRs.
https://github.com/pachi81/GlucoDataHandler/invitations

File: common/src/main/java/de/michelinside/glucodatahandler/common/receiver/NamedBroadcastReceiver.kt (hunk 1 above)
Comment (rgodha24) [reply common/src/main/java/de/michelinside/glucodatahandler/common/receiver/NamedBroadcastReceiver.kt:8]: Just to confirm, do you want me to change it back?

Comment (pachi81) [reply common/src/main/java/de/michelinside/glucodatahandler/common/receiver/NamedBroadcastReceiver.kt:8]: No, it is ok.

**Comment by rgodha24** [comment]
> I have also thought about making a small app, only doing this stuff and show all parts of the notification in main view. So users from other apps, like CampAPS can send me screenshots of the content extracted from the notification. What do you think?

I'm not sure I understand. Can you expand on this please?

> One more question, what about a packagename filter? For only handling notification from this package!?

This doesn't seem possible? at least I can't find any documentation on doing this in any other way than an if check in the NotificationListenerService

**Comment by pachi81** [comment]
> > I have also thought about making a small app, only doing this stuff and show all parts of the notification in main view. So users from other apps, like CampAPS can send me screenshots of the content extracted from the notification. What do you think?
> 
> I'm not sure I understand. Can you expand on this please?
> 
> > One more question, what about a packagename filter? For only handling notification from this package!?
> 
> This doesn't seem possible? at least I can't find any documentation on doing this in any other way than an if check in the NotificationListenerService
> 
> Thank you so much for the code review btw! I'm going to add support for Omnipod 5 IOB input later today then I'll mark it ready for review!

Sorry for the late response as I there was an emergency the whole weekend, after Abbott has changed there API. But it is working again, now....

> I'm not sure I understand. Can you expand on this please?

I mean, may to create a demo app with a button, which extracts the content of a notification in table format: name -> value
So I can give this app to an user to test its notification. Just an idea to handle new notifications of other apps.

File: common/src/main/java/de/michelinside/glucodatahandler/common/receiver/NotificationReceiver.kt [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    private val LOG_ID = "GDH.NotificationReceiver"
+
+    override fun getName(): String {
+        return LOG_ID
+    }
+
+    override fun onNotificationPosted(statusBarNotification: StatusBarNotification?) {
+        if (isRegistered()) {
+            statusBarNotification?.let { sbn ->
+                if (sbn.packageName == "com.dexcom.g7") {
```
Comment (pachi81) [inline common/src/main/java/de/michelinside/glucodatahandler/common/receiver/NotificationReceiver.kt:27]: As this is not only related to G7, I think it makes more sense to create a settings, selecting the package for parsing notification and compare name here.
But this is something, I can do, too.

File: common/src/main/java/de/michelinside/glucodatahandler/common/receiver/NotificationReceiver.kt [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+                    val bigContentView = notification.bigContentView
+
+                    // Try processing different RemoteViews
+                    if (processRemoteViews(contentView)) return
+                    if (processRemoteViews(bigContentView)) return
+
+                    // If we get here, we couldn't find the glucose value
+                    Log.d(LOG_ID, "Could not find glucose value in notification")
+                }
+                if (sbn.packageName == "com.insulet.myblue.pdm") {
```
Comment (pachi81) [inline common/src/main/java/de/michelinside/glucodatahandler/common/receiver/NotificationReceiver.kt:40]: Do you have any screenshot? Makes it easier for me to understand.
Is there also the glucose value present?
At the moment, 2 notifications are parsed, I would prefer only one for one app. 
But I do not have the knowledge for all these systems...

File: common/src/main/java/de/michelinside/glucodatahandler/common/receiver/NotificationReceiver.kt (hunk 5 above)
Comment (rgodha24) [reply common/src/main/java/de/michelinside/glucodatahandler/common/receiver/NotificationReceiver.kt:27]: Learning android dev is fun! I can try to figure this out in a few hours!

Comment (pachi81) [reply common/src/main/java/de/michelinside/glucodatahandler/common/receiver/NotificationReceiver.kt:27]: As you like!
I also started learning with this app, so maybe there are a lot of things to improve or which can be done in a better way...

Comment (rgodha24) [reply common/src/main/java/de/michelinside/glucodatahandler/common/receiver/NotificationReceiver.kt:27]: hey I've tried to get this working a few times on a few different days and I couldn't figure it out. If you still want this feature I unfortunately wont be able to finish it. my work is on https://github.com/rgodha24/GlucoDataHandler/tree/notification-receiver-chooser if you want

Comment (pachi81) [reply common/src/main/java/de/michelinside/glucodatahandler/common/receiver/NotificationReceiver.kt:27]: No problem, I can do it.

**Comment by ojalaj** [comment]
Thanks for the great app - thanks for the author and all the contributors.

I'm about to get OP5 & FSL2+. It seems that Glooko app is used to handle data.

As a former FSL2/FSL3 user that also uses WearOS watch and Android Auto, I'm now a bit worried about how to get CGM data in future. To my understanding, sensor readings are not transferred to Libreview cloud (since there is only one BT channel in FSL2+ sensor or something?).

Could Glugodatahandler do this job? Gluroo app seems to have already some solution for this.

**Comment by pachi81** [comment]
> Thanks for the great app - thanks for the author and all the contributors.
> 
> I'm about to get OP5 & FSL2+. It seems that Glooko app is used to handle data.
> 
> As a former FSL2/FSL3 user that also uses WearOS watch and Android Auto, I'm now a bit worried about how to get CGM data in future. To my understanding, sensor readings are not transferred to Libreview cloud (since there is only one BT channel in FSL2+ sensor or something?).
> 
> Could Glugodatahandler do this job? Gluroo app seems to have already some solution for this.

@ojalaj At the moment GDH can not read the notification, as it can not calculate the trend (which is not readable). So I will need to add historical data + graph to GDH (I´m working at the moment on this), then I need to find a formular to calculate the trend and if this is done, I will merge this change to GDH and enable notification read.
In the meantime you have to use xDrip+ in companion app mode to read the data from notification and forward to GDH.

**Comment by ojalaj** [comment]
> > Thanks for the great app - thanks for the author and all the contributors.
> > I'm about to get OP5 & FSL2+. It seems that Glooko app is used to handle data.
> > As a former FSL2/FSL3 user that also uses WearOS watch and Android Auto, I'm now a bit worried about how to get CGM data in future. To my understanding, sensor readings are not transferred to Libreview cloud (since there is only one BT channel in FSL2+ sensor or something?).
> > Could Glugodatahandler do this job? Gluroo app seems to have already some solution for this.
> 
> @ojalaj At the moment GDH can not read the notification, as it can not calculate the trend (which is not readable). So I will need to add historical data + graph to GDH (I´m working at the moment on this), then I need to find a formular to calculate the trend and if this is done, I will merge this change to GDH and enable notification read. In the meantime you have to use xDrip+ in companion app mode to read the data from notification and forward to GDH.

Thanks for the prompt reply @pachi81 ! This is not GDH related issue, but you think that (for now) xDrip+ is able to do the job with OP5&FSL2+ combo?

**Comment by pachi81** [comment]
> > > Thanks for the great app - thanks for the author and all the contributors.
> > > I'm about to get OP5 & FSL2+. It seems that Glooko app is used to handle data.
> > > As a former FSL2/FSL3 user that also uses WearOS watch and Android Auto, I'm now a bit worried about how to get CGM data in future. To my understanding, sensor readings are not transferred to Libreview cloud (since there is only one BT channel in FSL2+ sensor or something?).
> > > Could Glugodatahandler do this job? Gluroo app seems to have already some solution for this.
> > 
> > 
> > @ojalaj At the moment GDH can not read the notification, as it can not calculate the trend (which is not readable). So I will need to add historical data + graph to GDH (I´m working at the moment on this), then I need to find a formular to calculate the trend and if this is done, I will merge this change to GDH and enable notification read. In the meantime you have to use xDrip+ in companion app mode to read the data from notification and forward to GDH.
> 
> Thanks for the prompt reply @pachi81 ! This is not GDH related issue, but you think that (for now) xDrip+ is able to do the job with OP5&FSL2+ combo?

I think it is only related to OP5 and as @rgodha24 has developed this especially for OP5, it should work independent of the sensor you are using. So as long there is a notification xDrip+ can read the data and in some time also GDH will be able to read it.

**Comment by rgodha24** [comment]
This pr (right now) gets IOB data from the omnipod notification, and blood sugar data from the dexcom g7 notification. If you just have omnipod 5, it'll just get IOB data. I'm not a libre user, but I think xdrip+ has very good support for the libre2 so the companion mode idea that @pachi81 had should definitely work!

**Comment by pachi81** [comment]
> This pr (right now) gets IOB data from the omnipod notification, and blood sugar data from the dexcom g7 notification. If you just have omnipod 5, it'll just get IOB data. I'm not a libre user, but I think xdrip+ has very good support for the libre2 so the companion mode idea that @pachi81 had should definitely work!

Is there no notification of omnipad 5, which shows the glucose value? How does omnipad 5 works? I thought it will connect direct to the sensor?

**Comment by rgodha24** [comment]
The omnipod 5 does connect to the CGM, but its notification only contains IOB data. 
Screenshot 2025-01-28 at 11 05 18 AM
this is what the notification looks like.

**Comment by ojalaj** [comment]
To me it seems that Libre2+ is not used with the Libre App, but it is used with OP5 app. So you think that xDrip+ in Companion mode, when OP5 App is selected, will provide CGM data? Of course all this will be easier to test, when I get the OP5 in couple of weeks time, but I thought to prepare things for that beforehand.

**Comment by pachi81** [comment]
> To me it seems that Libre2+ is not used with the Libre App, but it is used with OP5 app. So you think that xDrip+ in Companion mode, when OP5 App is selected, will provide CGM data? Of course all this will be easier to test, when I get the OP5 in couple of weeks time, but I thought to prepare things for that beforehand.

Hard to say, if the OP5 only provide IOB in its complication, it will not work with xDrip companion mode for CGM data.
But as I do not know this system, better ask in a xDrip forum or OP5 forum.

**Comment by pachi81** [comment]
@rgodha24 just a question, is it also possible to read the trend direction from notification? Should be possible, if it is an unicode character!?

**Comment by rgodha24** [comment]
I'm not sure. I will look into it today and get back to you!

**Comment by rgodha24** [comment]
@pachi81 I did some research over the past 2 days, and because the dexcom g7 notifications are "remote content views", its really hard to parse data out of them. The arrows are images, not unicode, and I couldn't figure out how to get the image out of the notification class. 

Claude mentioned a possible solution with an accessibility class that I will look into this weekend.

**Comment by pachi81** [comment]
@rgodha24 I'd like to merge your changes to my development branch this weekend. Is it ok or are you still doing something?

**Comment by rgodha24** [comment]
@pachi81 I can clean it up and fix the merge conflicts! which branch are you planning on merging into?

**Comment by pachi81** [comment]
The chart branch. 
Thank you for your help!

**Comment by rgodha24** [comment]
@pachi81 merge conflicts are fixed! if you dont mind, can you do a squash and merge instead of a normal merge when you're pulling it into your branch?

File: auto/src/second/res/mipmap-anydpi-v26/ic_launcher.xml
Comment (pachi81) [inline auto/src/second/res/mipmap-anydpi-v26/ic_launcher.xml:1]: Can you check the removal of these files

Comment (rgodha24) [reply auto/src/second/res/mipmap-anydpi-v26/ic_launcher.xml:1]: not sure why they got deleted but I fixed it

**Comment by pachi81** [comment]
Looks, good, I will merge later. Which permission is needed?
